src/runtime/pkg/magent/metrics.go

```diff
@@ -0,0 +1,297 @@
+		wg.Add(1)
+		go func(sandboxID, namespace string, results chan<- []*dto.MetricFamily) {
+			sandboxMetrics, err := ma.getSandboxMetrics(sandboxID, namespace)
+			if err != nil {
+				logrus.WithError(err).Errorf("failed to get metrics for sandbox: %s", sandboxID)
+			}
+
+			results <- sandboxMetrics
+			wg.Done()
+			logrus.WithField("sandbox_id", sandboxID).Debug("job finished")
+		}(sandboxID, namespace, results)
+
+		logrus.WithField("sandbox_id", sandboxID).Debug("job started")
+	}
+
+	wg.Wait()
+	logrus.Debug("all job finished")
+	close(results)
+
+	// get all job result from chan
+	for sandboxMetrics := range results {
+		if sandboxMetrics != nil {
+			sandboxMetricsList = append(sandboxMetricsList, sandboxMetrics)
+		}
+	}
+
+	if len(sandboxMetricsList) == 0 {
+		return nil
+	}
+
+	// metricsMap used to aggregate metirc from multiple sandboxes
+	// key is MetricFamily.Name, and value is list of MetricFamily from multiple sandboxes
+	metricsMap := make(map[string]*dto.MetricFamily)
+	// merge MetricFamily list for the same MetricFamily.Name from multiple sandboxes.
+	for i := range sandboxMetricsList {
+		sandboxMetrics := sandboxMetricsList[i]
+		for j := range sandboxMetrics {
+			mf := sandboxMetrics[j]
+			key := *mf.Name
+
+			// add MetricFamily.Metric to the exists MetricFamily instance
+			if oldmf, found := metricsMap[key]; found {
+				oldmf.Metric = append(oldmf.Metric, mf.Metric...)
+			} else {
+				metricsMap[key] = mf
+			}
+		}
+	}
+
+	// write metrics to response.
+	for _, mf := range metricsMap {
+		if err := encoder.Encode(mf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getSandboxMetrics will get sandbox's metrics from shim
+func (ma *MAgent) getSandboxMetrics(sandboxID, namespace string) ([]*dto.MetricFamily, error) {
+	socket, err := ma.getMetricsAddress(sandboxID, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	transport := &http.Transport{
+		DisableKeepAlives: true,
+		Dial: func(proto, addr string) (conn net.Conn, err error) {
+			return net.Dial("unix", "\x00"+socket)
+		},
+	}
+
+	client := http.Client{
+		Timeout:   3 * time.Second,
+		Transport: transport,
+	}
+
+	resp, err := client.Get("http://shim/metrics")
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		resp.Body.Close()
+	}()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	reader := bytes.NewReader(body)
+	decoder := expfmt.NewDecoder(reader, expfmt.FmtText)
+
+	// decode metrics from sandbox to MetricFamily
+	list := make([]*dto.MetricFamily, 0)
+	for {
+		mf := &dto.MetricFamily{}
+		if err := decoder.Decode(mf); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		list = append(list, mf)
+	}
+
+	// newList contains processed MetricFamily
+	newList := make([]*dto.MetricFamily, len(list))
+
+	for i := range list {
+		metricFamily := list[i]
+		metricList := metricFamily.Metric
+		for j := range metricList {
+			metric := metricList[j]
```

Comment:
I wonder if we can avoid this loop and have it appended from the source instead?

Reply:
Yes, this loop can be reduced, and the processes is moved into loops above.